### PR TITLE
wait on yamux session close

### PR DESCRIPTION
**Pull request by bergwolf**
For both yamux client and server, we should wait on session close to get a clean state after close. Otherwise there might be inflight requests dirtying the serial channel and cause reconnection failures.

Fixes: #337

### CODE REVIEW

**Comment by jodh-intel** [comment]
Hi @bergwolf - the changes look good - any possibility of adding a unit test for this?

File: Gopkg.lock [hunk 1]
```diff
@@ -3,122 +3,95 @@
 
 [[projects]]
   branch = "master"
-  digest = "1:86feeb6c372fd7984322882737b00f32d7e75c6b437f93d75f43db3195574dab"
```
Comment (devimc) [inline Gopkg.lock]: seem like you are using an old version of `dep`

**Comment by bergwolf** [comment]
@devimc @jodh-intel PR updated. PTAL.

**Comment by jodh-intel** [comment]
Thanks @bergwolf.

lgtm

**Comment by jodh-intel** [comment]
From 16.04 initrd CI run:

```
=== RUN   TestNewAgentClientWithYamux
coverage: 81.9% of statements
panic: test timed out after 10s
```

Restarting...

File: protocols/client/client.go [hunk 2]
```diff
@@ -31,6 +32,7 @@ const (
 )
 
 var defaultDialTimeout = 15 * time.Second
+var defaultCloseTimeout = 5 * time.Second
```
Comment (devimc) [inline protocols/client/client.go]: have you tested this change with vsocks? see https://github.com/kata-containers/agent/issues/324 and https://github.com/kata-containers/agent/commit/93fe84e041b8895e68d268246a06c5a9a1395f3c

Comment (devimc) [reply protocols/client/client.go]: why you are reducing the timeout again?

Comment (bergwolf) [reply protocols/client/client.go]: This is close timeout not dial timeout. Why would a guest take more than 5 seconds to close a connection?

And it is only used by the yamux client. So only kata-runtime and kata-proxy will use it.

Comment (devimc) [reply protocols/client/client.go]: make sense, thanks

File: channel.go [hunk 3]
```diff
@@ -200,11 +201,16 @@ func (c *serialChannel) listen() (net.Listener, error) {
...
 	}
+	c.waitCh = session.CloseChan()
 
 	return session, nil
 }
 
 func (c *serialChannel) teardown() error {
+	// wait for the session to be fully shutdown first
+	if c.waitCh != nil {
+		<-c.waitCh
```
Comment (sboeuf) [inline channel.go]: You should use a go `select:` here, and timeout after some time, otherwise this might result into a deadlock.

Comment (bergwolf) [reply channel.go]: hmm, good point. I'll get to it later today.

**Pull request merged**